🧪 Add end-to-end tests for navigation

rye opened the pull request:
I like testing.

Testing is good.

"We should test more stuff" has been a long-standing thing I've said, but here I'm actually trying my hand at e2e testing.

I'm Not Sure™ if this affects coverage at all. It'd be nice if it did, but I don't think Detox instruments that yet, or if they do, I don't think we're uploading that. So, I guess I just have to give this a try.

***

list of views to test moved to #6788

rye committed:
Add a few testIDs to some things

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Add some e2e tests for navigation into the streaming view

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Sort props alphabetically, apparently

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

hawkrives commented on e2e/streaming.spec.js:
```diff
@@ -0,0 +1,41 @@
+		await expect(element(by.id('screen-homescreen'))).toNotBeVisible()
+		await expect(element(by.id('screen-streaming'))).toBeVisible()
+
+		await element(by.text('KSTO')).tap()
```
I think there’s a way to target tab bar items explicitly

rye replied:
Ah yes, after ~10 minutes of digging: [`tabBarTestID`](https://reactnavigation.org/docs/en/material-bottom-tab-navigator.html#tabbartestid) looks like the ticket. We will set that in the `navigationOptions` under the screens in such things in addition to the e.g. `tabBarLabel` and `tabBarIcon` things.

rye committed:
Replace deprecated relaunchApp call with launchApp call

Also, instead of deleting and reinstalling the app before each
individual test, (which is overkill) just launch a new instance of the
app.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Use tabBarTestID to find buttons instead of text search

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Try to fix an issue with not getting testIDs found

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Rename HomeScreenButton props and support testID on inner view

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
s/toNotBe/toBeNot/g

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
streaming: Remove assertion of screen-streaming

Since screen-streaming is actually just a TabNavigator, and I can't
quite figure out how to set the testID on it, I'm going to ignore it
and see if I can get a passing build

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Add testID prop support to RadioControllerView

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Merge branch 'master' into more-testing

I am doing this primarily to get the up-to-date GitHub Actions workflows
and such, but also the changes that have been made to various views.

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
e2e: Break apart the Streaming tests a bit more

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
e2e: Also test the KRLX Radio view

Signed-off-by: Kristofer Rye <kristofer.rye@gmail.com>

rye committed:
Merge branch 'master' into more-testing

hawkrives committed:
Merge branch 'master' into more-testing

* master: (2979 commits)
  update Podfile.lock
  update theme import and change names
  use standard "black" color
  switch codecov from the bash uploader to their action
  add new "dependency review" workflow
  add whitespace to "check" workflow jobs
  make the Prettier job's ID match its name
  upgrade linux runners to ubuntu-22.04
  centralize global versions for languages and Xcode
  auto-cancel in-progress jobs on subsequent commit to the same branch
  add distributionSha256Sum to gradle-wrapper.properties
  implement auto-cache-cleanup workflow
  Bump react-native-paper from 4.12.4 to 5.1.4
  Bump babel-jest from 29.3.1 to 29.4.0
  Bump @jest/globals from 29.3.1 to 29.4.0
  Bump @typescript-eslint/parser from 5.48.2 to 5.49.0
  Bump react-native-network-logger from 1.13.0 to 1.14.0
  Bump @typescript-eslint/eslint-plugin from 5.48.2 to 5.49.0
  Bump jest from 29.3.1 to 29.4.0
  Bump @react-native-community/datetimepicker from 6.7.2 to 6.7.3
  ...

hawkrives committed:
fix ts error in homescreen button

hawkrives committed:
fix missed merge conflict

hawkrives committed:
don't forget to launch the app!

hawkrives committed:
clean up implementation of homescreen buttons

hawkrives committed:
fix test for expected button targets

hawkrives committed:
remove deprecated "toBeNotVisible" method

hawkrives committed:
wait for the tab bar to finish appearing before tapping the tabs

hawkrives committed:
add test to open the Webcams screen

hawkrives committed:
simplify webcam button component

hawkrives committed:
rework streaming-media tests to run by user-facing text, instead of element ID

hawkrives committed:
remove more id matchers

hawkrives committed:
remove an ID matcher from the basic test

hawkrives committed:
port changes from this PR to the TS files

hawkrives committed:
remove test-ids from the homescreen buttons

rye reviewed (commented):
I can't ✅ this but I love it! This certainly seems to be going in the right direction.

As to overall strategy, I do think it's worth us exploring either mocking or full parametrization of things so that we have more control over data we expect to see in the app from the test suite and so that we can verify more tightly. However, I think we should do that in a separate MR — I think that's part of the larger infrastructurification push.
